---
![](images/spam-lord.png)

---
Harvesting emails and phone numbers
---

Here's your chance to be a SpamLord! Yes, you too can build regexes to help spread evil throughout the galaxy!

More specifically, your goal in this assignment is to write regular expressions that extract phone numbers and regular expressions that extract email addresses.

To start with a trivial example, given

    bspiering@usfca.edu
    
your job is to return

    bspiering@usfca.edu
    
    
![](https://imgs.xkcd.com/comics/hacking.png)

But you also need to deal with more complex examples created by people trying to shield their addresses, such as the following types of examples that I'm sure you've all seen:

    bspiering(at)usfca.edu
    bspiering at usfca dot edu

You should even handle examples that look like the following:

    <script type="text/javascript">obfuscate('usfca.edu','bspiering')</script> 

For all of the above you should return the corresponding email address

    brian.spiering@galvanize.com

as appropriate.  
Similarly, for phone numbers, you need to handle examples like the following:

    TEL +1-415-805-1888
    Phone:  (415) 805-1888
    Tel (+1): 415-805-1888
    <a href="contact.html">TEL</a> +1&thinsp;415&thinsp;805&thinsp;1888

all of which should return the following canonical form:

    415-805-1888

(you can assume all phone numbers we give you will be inside North America).  
In order to make it easier for you to do this, we will be giving you some data to test your code on, what is technically called a *development test set*. This is a document with some emails and phone numbers, together with the correct answers, so that you can make sure your code extracts all and only the right information. However, because you want to be sure your code *generalizes* well, you should be creative in looking at the web and thinking of different types of ways of encoding emails and phone numbers, and not just rely on the methods we've listed here.

**You won't have to deal with:** really difficult examples like images of any kind, or examples that require parsing names into first/last like:

    "first name"@usfca.edu

or difficult examples like:

    To send me email, try the simplest address that makes sense.

----
Where can I find the starter code?
----

You will find starter code in the github repository.  
By default, if you execute:

In [1]:
run spam_lord.py data/dev/ data/devGOLD

True Positives (4): 
{('balaji', 'e', 'balaji@stanford.edu'),
 ('nass', 'e', 'nass@stanford.edu'),
 ('shoham', 'e', 'shoham@stanford.edu'),
 ('thm', 'e', 'pkrokel@stanford.edu')}
False Positives (1): 
{('psyoung', 'e', 'young@stanford.edu')}
False Negatives (113): 
{('ashishg', 'e', 'ashishg@stanford.edu'),
 ('ashishg', 'e', 'rozm@stanford.edu'),
 ('ashishg', 'p', '650-723-1614'),
 ('ashishg', 'p', '650-723-4173'),
 ('ashishg', 'p', '650-814-1478'),
 ('bgirod', 'p', '650-723-4539'),
 ('bgirod', 'p', '650-724-3648'),
 ('bgirod', 'p', '650-724-6354'),
 ('cheriton', 'e', 'cheriton@cs.stanford.edu'),
 ('cheriton', 'e', 'uma@cs.stanford.edu'),
 ('cheriton', 'p', '650-723-1131'),
 ('cheriton', 'p', '650-725-3726'),
 ('dabo', 'e', 'dabo@cs.stanford.edu'),
 ('dabo', 'p', '650-725-3897'),
 ('dabo', 'p', '650-725-4671'),
 ('dlwh', 'e', 'dlwh@stanford.edu'),
 ('engler', 'e', 'engler@lcs.mit.edu'),
 ('engler', 'e', 'engler@stanford.edu'),
 ('eroberts', 'e', 'eroberts@cs.stanford.edu'),
 ('eroberts

---

You can also run the code at the command line:  


In [2]:
%%bash
python spam_lord.py data/dev/ data/devGOLD

True Positives (4): 
{('balaji', 'e', 'balaji@stanford.edu'),
 ('nass', 'e', 'nass@stanford.edu'),
 ('shoham', 'e', 'shoham@stanford.edu'),
 ('thm', 'e', 'pkrokel@stanford.edu')}
False Positives (1): 
{('psyoung', 'e', 'young@stanford.edu')}
False Negatives (113): 
{('ashishg', 'e', 'ashishg@stanford.edu'),
 ('ashishg', 'e', 'rozm@stanford.edu'),
 ('ashishg', 'p', '650-723-1614'),
 ('ashishg', 'p', '650-723-4173'),
 ('ashishg', 'p', '650-814-1478'),
 ('bgirod', 'p', '650-723-4539'),
 ('bgirod', 'p', '650-724-3648'),
 ('bgirod', 'p', '650-724-6354'),
 ('cheriton', 'e', 'cheriton@cs.stanford.edu'),
 ('cheriton', 'e', 'uma@cs.stanford.edu'),
 ('cheriton', 'p', '650-723-1131'),
 ('cheriton', 'p', '650-725-3726'),
 ('dabo', 'e', 'dabo@cs.stanford.edu'),
 ('dabo', 'p', '650-725-3897'),
 ('dabo', 'p', '650-725-4671'),
 ('dlwh', 'e', 'dlwh@stanford.edu'),
 ('engler', 'e', 'engler@lcs.mit.edu'),
 ('engler', 'e', 'engler@stanford.edu'),
 ('eroberts', 'e', 'eroberts@cs.stanford.edu'),
 ('eroberts

----
It will run your code on the files contained in data/dev/ and compare the results of a simple regular expression against the correct results. The results will look something like this:

    True Positives (4): 
    set([('mike', 'e', 'mike@stanford.edu'),
         ('nass', 'e', 'nass@stanford.edu'),
         ('shoham', 'e', 'shoham@stanford.edu'),
         ('thm', 'e', 'pkrokel@stanford.edu')])
    False Positives (1): 
    set([('psyoung', 'e', 'young@stanford.edu')])
    False Negatives (113): 
    set([('ashishg', 'e', 'ashishg@stanford.edu'),
         ('ashishg', 'e', 'rozm@stanford.edu'),
         ('ashishg', 'p', '650-723-1614'),
    ...

The format is: (`file_name`, `(e)mail or (p)hone`, `correct_answer`)

The __true positive__ section displays e-mails and phone numbers which the starter code correctly matches.

The __false positive__ section displays e-mails which the starter code regular expressions match but which are not correct

The __false negative__ section displays e-mails and phone numbers which the starter code did not match, but which do exist in the html files. Go the file and see why your regex pattern is not matching the email or phone

Your goal, then, is to reduce the number of false positives and negatives to 0️⃣. 

----
Check for Understanding
-----
<br>
<details><summary>
Why do you not care about true negatives?
</summary>
There are too many TN. It is all the text that is not a phone number or email!
</details>

Where should I write my Python code?
----

The function
```python
def process_file(name, f):
```
has been flagged for you with the universal "TODO" marker. This function takes in a file object (or any iterable of strings) and returns a list of tuples representing e-mails or phone numbers found in that file. Specifically the tuple has the format:

    (filename, type, value)

where type is either 'e', for e-mail, or 'p' for phone number, and value is just the actual e-mail or phone number.

What format should the final phone numbers and e-mail have?
---

The canonical forms expected are:

    user@example.com
    415-555-1234

The case of the e-mails you find should not matter because the starter code will lowercase your matched e-mails before comparing them against the gold set

----
Hints
-----

- Look at the raw data in `./data/dev/`. They are saved webpages.
- Look at the correct answers in `./data/devGOLD`
- Be careful about online Python regex checkers. They tend to use Python 2 which is different.
- Start with phone numbers then tackle emails. 
- It is possible to get to 💯

---
When my regex works on first try
----

![](http://tclhost.com/vngeKjr.gif)

[Source](http://thecodinglove.com/post/128186083300/when-my-regex-works-on-first-try)

<br>
<br>